Clone dirhandles without fchdir

Leont opened the pull request:
This is a WIP for fixing #23010 based on @ilmari's suggestion. There currently are two problems with it.

1. It doesn't have a Configure check yet for `fdopendir`
2. `t/op/threads-dirh.t` fails because it's assuming that the dirhandles in different threads are independent. Given that the same isn't true of filehandles, I think this is the wrong thing to aim at anyway.

Leont committed:
Clone dirhandles without fchdir

This uses fdopendir and dup to dirhandles. Unlike the fchdir approach
these handles share descriptor position among threads.

tonycoz commented:
I think it's one reasonable fix, though I think it would be an entry in incompatible changes in perldelta.

I'm not sure it's safe for multiple threads to readdir() from the same underlying fd: It's UB for both parent and child of a fork() to read from their corresponding copies of the `DIR`, which is the closest case I can think of to this, so I'm not sure this will be thread safe.

From my testing with simple test code, glibc gets confused about the telldir() position for the new handle.

One option might be to `openat(orig_dir_fd, ".", O_DIRECTORY | ...)` which, assuming no permission change on `.`, should give us an independent handle, which can be positioned like we do in the current code.

Of course, the simplest option is to just not pass the `DIR *` into the new interpreter, which resolves the original issue in #10387, but is as backward (in)compatible as this change.

Leont commented:
> I'm not sure it's safe for multiple threads to readdir() from the same underlying fd: It's UB for both parent and child of a fork() to read from their corresponding copies of the DIR, which is the closest case I can think of to this, so I'm not sure this will be thread safe.

It's not, but the same is true across forks. The same is also true across threads for filehandles, really. This all falls firmly in "doctor it hurts when I poke into my eyeballs" IMO.

ap commented:
Any movement here?

Leont commented:
> Any movement here?

I suspect I know how to write the missing parts in a fairly short time, but I'm not sure I'd be comfortable including it this late in the release cycle

ap commented:
Note that @vinc17fr has just [posted to oss-security](https://www.openwall.com/lists/oss-security/2025/05/22/2) regarding #23010.

Leont committed:
Detect presence of fdopendir in Configure

Leont committed:
Remove dubious thread dirhandle cloning tests

These tests makes assumptions that are only true for the current fchdir
based implementation of dirhandle cloning. In particular that the cloned
dirhandle is entirely separate from the original. This is not
appropriate for the upcoming fdopendir based implementation.

jkeenan commented on Porting/Glossary:
```diff
@@ -947,6 +947,11 @@ d_fmin (d_fmin.U):
 
+d_fdopendir (d_fdopendir.U):
+	This variable conditionally defines the HAS_FORK symbol, which
+	indicates that the fdopen routine is * available to open a
```
The asterisk in `is * available` appears to be extraneous.

Leont replied:
Oops, yeah I should fix that.

jkeenan commented on t/op/threads-dirh.t:
```diff
@@ -31,101 +27,3 @@ fresh_perl_is <<'# this is no comment', 'ok', {}, 'crash when duping dirh';
- is scalar @w, 0, 'no warnings during all that' or diag @w;
- chdir updir;
-}
-rmtree($dir);
```
Can you briefly explain why we should remove this somewhat large block of test code?  How do we exercise the code needed to satisfy the problems discussed in GH #23010?  Thanks.

Leont replied:
That's item 2 in the ticket description. But yeah I should explain that in the commit too.

khwilliamson commented on sv.c:
```diff
@@ -14106,6 +14106,7 @@ Perl_dirp_dup(pTHX_ DIR *const dp, CLONE_PARAMS *const param)
 #endif
 
     PERL_UNUSED_CONTEXT;
+    PERL_UNUSED_VAR(param);
```
technically, shouldn't this be  `PERL_UNUSED_ARG`

Leont replied:
I'm not sure why that distinction even exists, but I'll change it.

khwilliamson commented on win32/config.vc:
```diff
@@ -199,6 +199,7 @@ d_fd_macros='define'
 d_fd_set='define'
 d_fdclose='undef'
 d_fdim='undef'
+d_fdopendir='undef'
```
Should `HAS_FDOPENDIR`  be added to `metaconfig.h`

Leont replied:
I suspect not, but @Tux would know for sure

Leont replied:
As far as I can tell it only matters when regenerating Configure from metaconfig anyway. So I propose we leave it out for now and if it turns out to be necessary Tux can add it later when he comes back from his vacation.

Leont committed:
Remove fchdir based dirhandle cloning code

This is no longer needed now that the fdopendir implementation exists.

ap commented:
Given that Windows has its own logic for dirhandle dup’ing, that `fdopendir` is in POSIX same as `fchdir` is (and apparently `fdopendir` is actually the older one), and that both appear to be widely supported in mainstream Unixes, it seems that deleting the `fchdir` implementation would not shift our platform support posture materially, if at all. Only some obscure platforms may be affected.

I’ve discussed with @book and we agreed to drop the `fchdir` logic. We haven’t been able to reach @haarg so maybe he’ll object, but please amend the patch accordingly so it’s ready to go if he agrees, since we’re already late on the point release. (If he objects it will be easy enough to just use the current state of the patch.)

haarg commented:
I concur that preserving the `fchdir` code path seems like a bad idea.
